File: sdks/typescript/src/events.ts
import type { StarfishFrame, EventFilter } from "./types.js";
import { EventStream, type Unsubscribe } from "./emitter.js";

export class Events {
  private stream = new EventStream<StarfishFrame>();

  dispatch(frame: StarfishFrame): void {
    this.stream.emit(frame);
  }

  events$(filter?: EventFilter): EventStream<StarfishFrame> {
    if (!filter) return this.stream;

    const filtered = new EventStream<StarfishFrame>();

    this.stream.subscribe((frame) => {
      if (filter.resource && frame.header.resource !== filter.resource) return;
      if (filter.method && frame.header.method !== filter.method) return;
      if (filter.topic && frame.header.topic !== filter.topic) return;
      if (filter.from && frame.header.from !== filter.from) return;
      filtered.emit(frame);
    });

    return filtered;
  }

  subscribe(callback: (frame: StarfishFrame) => void): Unsubscribe {
    return this.stream.subscribe(callback);
  }
}

File: sdks/swift/Sources/StarfishClient/Events.swift
import Foundation

/// Global event dispatching with optional filtering.
final class Events: @unchecked Sendable {
    private let stream = EventStream<StarfishFrame>()

    func dispatch(_ frame: StarfishFrame) {
        stream.emit(frame)
    }

    /// Returns an EventStream filtered by the given criteria.
    func events$(filter: EventFilter? = nil) -> EventStream<StarfishFrame> {
        guard let filter = filter else { return stream }

        let filtered = EventStream<StarfishFrame>()
        stream.subscribe { frame in
            if let resource = filter.resource, frame.header.resource != resource { return }
            if let method = filter.method, frame.header.method != method { return }
            if let topic = filter.topic, frame.header.topic != topic { return }
            if let from = filter.from, frame.header.from != from { return }
            filtered.emit(frame)
        }
        return filtered
    }

    /// Subscribe to all events with a callback.
    @discardableResult
    func subscribe(_ callback: @escaping @Sendable (StarfishFrame) -> Void) -> Unsubscribe {
        stream.subscribe(callback)
    }
}
